NF: clone: --reckless='ephemeral'

bpoldrack opened the pull request:
Introduce reckless mode 'ephemeral' for throw-away clones, symlinking their .git/annex to origin's .git/annex.
In addition declare 'here' dead to not propagate availability info from it when publishing captured changes back.

bpoldrack commented:
Currently not able to grasp what's going on in the windows test. Also it fails differently on our windows testing box. Any thoughts, @mih?

kyleam commented on datalad/core/distributed/clone.py:
```diff
@@ -571,16 +586,64 @@ def postclonecfg_annexdataset(ds, reckless, description=None):
 
+    elif reckless == 'ephemeral':
+        # with ephemeral we declare 'here' as 'dead' right away, whenever
+        # we symlink origins annex. Because we want annex to copy to
```
s/origins/origin's/

kyleam commented on datalad/core/distributed/clone.py:
```diff
@@ -571,16 +586,64 @@ def postclonecfg_annexdataset(ds, reckless, description=None):
+        # with ephemeral we declare 'here' as 'dead' right away, whenever
+        # we symlink origins annex. Because we want annex to copy to
+        # the ria remote to get availability info correct for an eventual
+        # git-push into the store
```
I'm having a hard time parsing/understanding from "Because [...]" to here.  If it is meant to convey what is said in the commit message (re: not propagating back), I think that phrasing is clearer.

kyleam commented on datalad/core/distributed/clone.py:
```diff
@@ -571,16 +586,64 @@ def postclonecfg_annexdataset(ds, reckless, description=None):
+            'remote.origin.annex-ignore', 'true',
+            where='local')
+
+        ds.repo._run_annex_command('dead', annex_options=['here'])
```
`ds.repo.set_remote_dead('here')`?

kyleam commented on datalad/core/distributed/clone.py:
```diff
@@ -571,16 +586,64 @@ def postclonecfg_annexdataset(ds, reckless, description=None):
+            if origin_annex_url:
+                try:
+                    # deal with file:// scheme URLs as well as plain paths
+                    # if origin isn't local, we have nothing to do
```
comment nit-pick: I'm frequently tripped up by comments in this project that drop a period and treat a new line as a sentence separator.  As a reader, I would prefer to not have to guess from the context.  Like here, I'd guess that it should be "paths. If", but I'm not sure.

kyleam commented on datalad/core/distributed/clone.py:
```diff
@@ -571,16 +586,64 @@ def postclonecfg_annexdataset(ds, reckless, description=None):
+                    origin_git_path = Path(PathRI(origin_annex_url).localpath)
+                    if origin_git_path.name != '.git':
+                        origin_git_path /= '.git'
+                except Exception:
```
Can a more specific exception be caught here?

kyleam commented on datalad/core/distributed/clone.py:
```diff
@@ -571,16 +586,64 @@ def postclonecfg_annexdataset(ds, reckless, description=None):
+                    if origin_git_path.name != '.git':
+                        origin_git_path /= '.git'
+                except Exception:
+                    # TODO: What level? + note, that annex-dead is independ
```
s/independ/independent/?  I'm not sure what the note part of the to-do is about.

kyleam commented on datalad/core/distributed/clone.py:
```diff
@@ -571,16 +586,64 @@ def postclonecfg_annexdataset(ds, reckless, description=None):
+                                     target_is_directory=True)
+        else:
+            # TODO: What level? + note, that annex-dead is independ
+            lgr.warning("Unable to create symlinks.")
```
I think something about "reckless=ephemeral" should be added to this message.

kyleam commented on datalad/core/distributed/tests/test_clone.py:
```diff
@@ -763,3 +765,72 @@ def test_ria_http_storedataladorg(path):
     eq_(ds.id, datalad_store_testds_id)
 
+
+@skip_if_on_windows  # see gh-
```
s/gh-/&4131/

bpoldrack commented:
Thanks, @kyleam - will address those comments.

>  Playing around with this a bit, I noticed that I could trigger a failure if I did anything in the original repository (the one the symlink points to). Have you hit into any issues on your end?

I can reproduce that, yes. However, I don't think we can do much about that, other than maybe enhancing the doc for it (and possibly complain to Joey). Annex complains about the object for the file you saved in origin, since in `clone1` it doesn't know about it. IMO this shouldn't necessarily lead to failure to commit on annex' end, since an unknown object in the object tree doesn't really hurt (particularly not the new object that we tried to commit).

So, in my (current) view, we should add to the doc, that a symlinked annex comes with some potential issues (it's called `reckless` after all). A `save` should prob. be prepended by an `update` ( `git fetch` actually is sufficient), which solves the case you set up. The test case would still fail on publishing back, since it's then not fast-forward, of course.

The things we possibly could  do, I think, are:

1. catch that error on `save`, detect whether `.git/annex` is symlinked and suggest to update first via error message OR

2. save `ephemeral` mode in repo (we prob. should anyway, since parameters to clone should be stored as the default for subdatasets - just noticed, that I forgot to make sure that's true for `reckless` as well) and based on that config have an automatic fetch prior to a save.

What do you think @mih ?

mih commented:
@kyleam Thanks for the review!

@bpoldrack your reasoning seems sound!

kyleam commented:
@bpoldrack  Thanks for the pushed updates.  Those look good to me.

kyleam commented:
For posterity:
  * original discussion around this idea: gh-3521
  * related issue that is partially fulfilled by this pr: gh-3528

bpoldrack commented:
Changed storing the `reckless` mode accordingly. And added to the doc.
From my point of view this is it for this PR. If you guys agree, @kyleam , @mih , I would make an issue regarding dealing with detecting the potential issue from within `save` and have a separate PR for that.

kyleam commented:
Sounds good to me.  Thanks @bpoldrack.

kyleam commented on datalad/core/distributed/clone.py:
```diff
@@ -652,6 +645,16 @@ def postclonecfg_annexdataset(ds, reckless, description=None):
+            'datalad.clone.reckless', reckless,
+            where='local',
+            # delay reload until all config IO is done
+            reload=False)
```
This lack of a reload is what's behind the new test failures.  In the deleted call above, the very next line has `reload=True`.

bpoldrack replied:
Thx!
